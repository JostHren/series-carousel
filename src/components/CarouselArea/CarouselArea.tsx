import { selectSeriesQuery } from "../../store/series/seriesSlice";
import { TvMazeSeries } from "../../store/services/types";
import { useGetSeriesByNameQuery } from "../../store/services/series";
import { useSelector } from "react-redux";
import { getRandomFromArray } from "../../helpers";
import { useMemo } from "react";
import { HeroCarousel } from "../HeroCarousel/HeroCarousel";
import { SeriesCarousel } from "../SeriesCarousel/SeriesCarousel";
import {
  LOADING,
  ERROR,
  NO_SEARCH_QUERY,
  NO_RESULTS,
  NOT_POPULAR,
} from "./constants";

export const CarouselArea = (): JSX.Element => {
  const query = useSelector(selectSeriesQuery());
  const { error, isLoading, data, isFetching } = useGetSeriesByNameQuery(
    query ?? ""
  );

  const threeRandomSeries: TvMazeSeries[] = useMemo(
    () => getRandomFromArray(data ?? [], 3),
    [data]
  );

  const restOfSeries = useMemo(
    () =>
      data?.filter((show) => {
        const checkForSameSeriesId = (series: TvMazeSeries): boolean =>
          series.show.id === show.show.id;

        return !threeRandomSeries.some(checkForSameSeriesId);
      }),
    [data, threeRandomSeries]
  );

  if (isFetching || isLoading) return <HeroCarousel data={LOADING} />;
  if (error) return <HeroCarousel data={ERROR} />;
  if (!query || query === "") return <HeroCarousel data={NO_SEARCH_QUERY} />;
  if (!data || data.length === 0) return <HeroCarousel data={NO_RESULTS} />;

  return (
    <div className="flex flex-col sm:flex-row justify-between min-w-fit">
      <HeroCarousel data={threeRandomSeries} />
      <SeriesCarousel
        data={
          !restOfSeries || restOfSeries?.length === 0
            ? NOT_POPULAR
            : restOfSeries
        }
      />
    </div>
  );
};
