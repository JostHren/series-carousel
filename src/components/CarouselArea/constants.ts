export const NO_SEARCH_QUERY = [
  {
    show: {
      id: 0o0,
      image: { original: "./query.png" },
    },
  },
];

export const LOADING = [
  {
    show: {
      id: 0o0,
      image: { original: "./loading.png" },
    },
  },
];

export const ERROR = [
  {
    show: {
      id: 0o0,
      image: { original: "./error.png" },
    },
  },
];

export const NO_RESULTS = [
  {
    show: {
      id: 0o0,
      image: { original: "./no_results.png" },
    },
  },
];

export const NOT_POPULAR = [
  {
    show: {
      id: 0o0,
      image: { original: "./not_popular.png" },
      name: "Try something else",
    },
  },
];
