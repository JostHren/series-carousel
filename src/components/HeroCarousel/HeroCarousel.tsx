import { useCarouselNavigation, useCarouselNavigationProps } from "../../hooks";
import { TvMazeSeries } from "../../store/services/types";

export interface CarouselProps {
  data: Partial<TvMazeSeries>[];
  slideInterval?: number;
  slidesPerPage?: number;
}

export interface CarouselSlideProps {
  slide: Partial<TvMazeSeries>;
}

const HeroCarouselLabel = (): JSX.Element => {
  return (
    <div className="absolute border-red-900 border-4 rounded-md bg-slate-900 p-2 text-red-800 text-lg font-extrabold left-5 top-3 -rotate-6 z-50 shadow-md">
      Random 3
    </div>
  );
};

const HeroCarouselNavigation = ({
  nextSlide,
  prevSlide,
}: Partial<useCarouselNavigationProps>): JSX.Element | null => {
  if (!nextSlide || !prevSlide) return null;

  return (
    <>
      <button
        className="absolute bg-slate-800 bg-opacity-0 hover:bg-opacity-30 top-0 bottom-0 left-0 z-10 w-20"
        onClick={() => prevSlide()}
      >
        <svg
          className="m-auto"
          width={"50%"}
          xmlns="http://www.w3.org/2000/svg"
          viewBox="0 0 24 24"
          fill="#F6E05E"
        >
          <path d="M15.71 20.71a1 1 0 0 0 0-1.41l-7.29-7.29 7.29-7.29a.996 .996 0 1 0-1.41-1.41l-8.01 7.98c-.2.2-.29.45-.29.71s.1.51.29.71l8 8c.39.39 1.02.39 1.41 0Z"></path>
        </svg>
      </button>
      <button
        className="absolute bg-slate-800 bg-opacity-0 hover:bg-opacity-30 top-0 bottom-0 right-0 w-20 z-10"
        onClick={() => nextSlide()}
      >
        <svg
          className="m-auto"
          width={"50%"}
          xmlns="http://www.w3.org/2000/svg"
          viewBox="0 0 24 24"
          fill="#F6E05E"
        >
          <path d="M8.29 3.29a1 1 0 0 0 0 1.41l7.29 7.29-7.29 7.29a.996.996 0 1 0 1.41 1.41l8-8c.2-.2.29-.45.29-.71s-.1-.51-.29-.71L9.71 3.29a1 1 0 0 0-1.41 0Z"></path>
        </svg>
      </button>
    </>
  );
};

const HeroCarouselSlide = ({ slide }: CarouselSlideProps): JSX.Element => {
  return (
    <div
      className="min-w-full h-[50vh] sm:h-[90vh] relative"
      key={slide.show?.id}
    >
      <div className="min-w-full relative">
        <div className="absolute bottom-0 m-2 text-yellow-400">
          <div className="bg-black text-lg font-extrabold mb-2 p-1 z-30 ">
            {slide.show?.name}
          </div>
          <div className="bg-black w-min truncate mb-1 p-1 z-30 ">
            {slide.score
              ? `Rating: ${Math.round(slide.score * 100) / 100}`
              : null}
          </div>
          <div className="bg-black w-min truncate p-1 z-30">
            {slide.show?.status ? `Status: ${slide.show.status}` : null}{" "}
          </div>
        </div>
        <div
          className="m-auto h-[50vh] sm:h-[90vh]"
          style={{
            backgroundImage: `url(${
              slide.show?.image?.original ?? "./not_found.png"
            })`,
            backgroundPosition: "center",
            backgroundSize: "contain",
            backgroundRepeat: "no-repeat",
          }}
        ></div>
      </div>
    </div>
  );
};

export const HeroCarousel = ({
  data,
  slideInterval = 3000,
  slidesPerPage = 1,
}: CarouselProps): JSX.Element => {
  const { isErrorSlide, currentIndex, nextSlide, prevSlide } =
    useCarouselNavigation({
      data,
      slideInterval,
      slidesPerPage,
    });

  return (
    <div className="sm:m-auto basis-full sm:basis-4/6 relative bg-black overflow-hidden">
      {!isErrorSlide && <HeroCarouselLabel />}
      <HeroCarouselNavigation prevSlide={prevSlide} nextSlide={nextSlide} />
      <div
        className="flex min-w-full transition-transform duration-500"
        style={{
          transform: `translateX(-${currentIndex * 100}%)`,
        }}
      >
        {data.map((slide) => (
          <HeroCarouselSlide slide={slide} key={slide.show?.id} />
        ))}
      </div>
    </div>
  );
};
