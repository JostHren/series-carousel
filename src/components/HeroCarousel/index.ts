export type { CarouselProps, CarouselSlideProps } from "./HeroCarousel";
export { HeroCarousel } from "./HeroCarousel";
