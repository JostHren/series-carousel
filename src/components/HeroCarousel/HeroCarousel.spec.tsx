import { render } from "@testing-library/react";

import { HeroCarousel } from "./HeroCarousel";
import { NOT_POPULAR } from "../CarouselArea/constants";

describe("HeroCarousel component", () => {
  it("HeroCarousel component to render", () => {
    const { container } = render(<HeroCarousel data={NOT_POPULAR} />);

    expect(container).toMatchSnapshot();
  });
});
