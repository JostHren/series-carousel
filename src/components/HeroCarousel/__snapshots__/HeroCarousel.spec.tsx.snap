// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HeroCarousel component > HeroCarousel component to render 1`] = `
<div>
  <div
    class="basis-full sm:basis-4/6 relative bg-black overflow-hidden max-h-[calc(61dvh-6rem)] sm:max-h-[calc(100dvh-4rem)]"
  >
    <button
      class="absolute bg-slate-800 bg-opacity-0 hover:bg-opacity-30 top-0 bottom-0 left-0 z-10 w-20"
    >
      <svg
        class="m-auto"
        fill="#F6E05E"
        viewBox="0 0 24 24"
        width="50%"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M15.71 20.71a1 1 0 0 0 0-1.41l-7.29-7.29 7.29-7.29a.996 .996 0 1 0-1.41-1.41l-8.01 7.98c-.2.2-.29.45-.29.71s.1.51.29.71l8 8c.39.39 1.02.39 1.41 0Z"
        />
      </svg>
    </button>
    <button
      class="absolute bg-slate-800 bg-opacity-0 hover:bg-opacity-30 top-0 bottom-0 right-0 w-20 z-10"
    >
      <svg
        class="m-auto"
        fill="#F6E05E"
        viewBox="0 0 24 24"
        width="50%"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8.29 3.29a1 1 0 0 0 0 1.41l7.29 7.29-7.29 7.29a.996.996 0 1 0 1.41 1.41l8-8c.2-.2.29-.45.29-.71s-.1-.51-.29-.71L9.71 3.29a1 1 0 0 0-1.41 0Z"
        />
      </svg>
    </button>
    <div
      class="flex min-w-full transition-transform duration-500"
      style="transform: translateX(-0%);"
    >
      <div
        class="min-w-full relative"
      >
        <div
          class="min-w-full relative"
        >
          <div
            class="absolute bottom-0 m-2 text-yellow-400"
          >
            <div
              class="bg-black text-lg font-extrabold mb-2 p-1 z-30 "
            >
              Try something else
            </div>
            <div
              class="bg-black w-min truncate mb-1 p-1 z-30 "
            />
            <div
              class="bg-black w-min truncate p-1 z-30"
            >
               
            </div>
          </div>
          <div
            class="m-auto h-[calc(61dvh-6rem)] sm:h-[calc(100dvh-4rem)]"
            style="background-image: url(./not_popular.png); background-position: center; background-size: contain; background-repeat: no-repeat;"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
